core/vault.go

```diff
@@ -0,0 +1,416 @@
+package core
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/big"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/starkandwayne/goutils/log"
+)
+
+type Vault struct {
+	URL            string
+	Token          string
+	EncryptionType string
+	Insecure       bool
+	HTTP           *http.Client
+}
+
+type VaultCreds struct {
+	SealKey        string `json:"seal_key"`
+	RootToken      string `json:"root_token"`
+	EncryptionType string `json:"encryption_type"`
+}
+
+var status struct {
+	Sealed bool `json:"sealed"`
+}
+
+func (vault *Vault) Init(store string) error {
+	initialized, err := vault.IsInitialized()
+	if err != nil {
+		return err
+	}
+
+	if initialized {
+		log.Infof("vault is already initialized")
+
+		reader := bufio.NewReader(os.Stdin)
+		fmt.Print("Enter Master Password: ")
+		master, _ := reader.ReadString('\n')
+		master = strings.TrimSpace(master)
+
+		creds, err := vault.ReadConfig(store, master)
+		if err != nil {
+			return err
+		}
+		vault.Token = creds.RootToken
+		vault.EncryptionType = creds.EncryptionType
+		os.Setenv("VAULT_TOKEN", vault.Token)
+		return vault.Unseal(creds.SealKey)
+	}
+
+	//////////////////////////////////////////
+
+	log.Infof("initializing the vault with 1/1 keys")
+	res, err := vault.Do("PUT", "/v1/sys/init", map[string]int{
+		"secret_shares":    1,
+		"secret_threshold": 1,
+	})
+	if err != nil {
+		log.Errorf("failed to initialize the vault: %s", err)
+		return err
+	}
+	b, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		log.Errorf("failed to read response from the vault, concerning our initialization attempt: %s", err)
+		return err
+	}
+
+	var keys struct {
+		RootToken string   `json:"root_token"`
+		Keys      []string `json:"keys"`
+	}
+	if err = json.Unmarshal(b, &keys); err != nil {
+		log.Errorf("failed to parse response from the vault, concerning our initialization attempt: %s", err)
+		return err
+	}
+	if keys.RootToken == "" || len(keys.Keys) != 1 {
+		if keys.RootToken == "" {
+			log.Errorf("failed to initialize vault: root token was blank")
+		}
+		if len(keys.Keys) != 1 {
+			log.Errorf("failed to initialize vault: incorrect number of seal keys (%d) returned", len(keys.Keys))
+		}
+		err = fmt.Errorf("invalid response from vault: token '%s' and %d keys", keys.RootToken, len(keys.Keys))
+		return err
+	}
+
+	reader := bufio.NewReader(os.Stdin)
+	fmt.Print("Enter Encryption Cipher (aes-128, aes-256, twofish): ")
```

Comment:
Encryption type should be configured by the administrator, in the core yaml config file (with a sane, best practices default)